compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -748,6 +748,11 @@ typeOf' = \case
     checkImplements tpl iface
     checkExpr val (TCon iface)
     pure (TOptional (TCon tpl))
+  EUnsafeFromInterface iface tpl cid val -> do
+    checkImplements tpl iface
+    checkExpr cid (TContractId (TCon tpl))
```

Comment:
Even though we primarily use this when we have (or think we have) a template contract id, I think it makes more sense for the primitive to take `TContractId (TCon iface)` instead of `TContractId (TCon tpl)`.